crates/steel/src/event.rs

```diff
@@ -0,0 +1,192 @@
+use std::marker::PhantomData;
+
+/// Represents an EVM event query.
+///
+/// ### Usage
+/// - **Preflight calls on the Host:** To prepare the event query on the host environment and build
+///   the necessary proof, use [Event::preflight].
+/// - **Calls in the Guest:** To initialize the event query in the guest, use [Event::new].
+///
+/// ### Examples
+/// ```rust,no_run
+/// # use risc0_steel::{ethereum::EthEvmEnv, Event};
+/// # use alloy_primitives::address;
+/// # use alloy_sol_types::sol;
+///
+/// # #[tokio::main(flavor = "current_thread")]
+/// # async fn main() -> anyhow::Result<()> {
+/// let contract_address = address!("dAC17F958D2ee523a2206206994597C13D831ec7");
+/// sol! {
+///     interface IERC20 {
+///         event Transfer(address indexed from, address indexed to, uint256 value);
+///     }
+/// }
+///
+/// // Host:
+/// let url = "https://ethereum-rpc.publicnode.com".parse()?;
+/// let mut env = EthEvmEnv::builder().rpc(url).build().await?;
+/// let event = Event::preflight::<IERC20::Transfer>(&mut env).address(contract_address);
+/// event.query().await?;
+///
+/// let evm_input = env.into_input().await?;
+///
+/// // Guest:
+/// let env = evm_input.into_env();
+/// let event = Event::new::<IERC20::Transfer>(&env).address(contract_address);
+/// let logs = event.query();
+///
+/// # Ok(())
+/// # }
+/// ```
+///
+/// [Contract]: crate::Contract
+pub struct Event<S, E> {
+    filter: Filter,
+    env: E,
+    phantom: PhantomData<S>,
+}
+
+impl<H: EvmBlockHeader> Event<(), &GuestEvmEnv<H>> {
+    /// Constructor for executing an event query for a specific Solidity event.
+    pub fn new<S: SolEvent>(env: &GuestEvmEnv<H>) -> Event<S, &GuestEvmEnv<H>> {
+        Event {
+            filter: event_filter::<S, H>(env.header()),
+            env,
+            phantom: PhantomData,
+        }
+    }
+}
+
+#[cfg(feature = "unstable-event")]
+impl<S: SolEvent, H: EvmBlockHeader> Event<S, &GuestEvmEnv<H>> {
+    /// Executes the query and returns the matching logs and panics on failure.
+    ///
+    /// A convenience wrapper for [Event::try_query], panicking if the call fails. Useful when
+    /// success is expected.
+    pub fn query(self) -> Vec<Log<S>> {
+        self.try_query().unwrap()
+    }
+
+    /// Attempts to execute the query and returns the matching logs or an error.
+    pub fn try_query(self) -> anyhow::Result<Vec<Log<S>>> {
+        let logs = WrapStateDb::new(self.env.db(), &self.env.header).logs(self.filter)?;
+        logs.iter()
+            .map(|log| Ok(S::decode_log(log, false)?))
+            .collect()
+    }
+}
+
+impl<S, E> Event<S, E> {
+    /// Sets the address to query with this filter.
+    ///
+    /// See [`Filter::address`].
+    pub fn address<A: Into<ValueOrArray<Address>>>(mut self, address: A) -> Self {
+        self.filter.address = address.into().into();
+        self
+    }
+
+    /// Sets the 1st indexed topic.
+    pub fn topic1<TO: Into<Topic>>(mut self, topic: TO) -> Self {
```

Comment:
Is is possible to additionally provide a more type-safe option using `SolEvent::TopicList`? It would be nice, but I can't tell from their docs how to make this work myself.